### DISCUSSION TITLE
uiPanelCanvasRectangle and uiPanelSnapSize tricks for VST restricted instances

### DISCUSSION THREAD

**Discussion by damiensellier** [RomanKubiak/ctrlr#240] [Q&A] [answered]
Hi,
Yesterday I had some trouble exporting my panel as a VST restricted instance because I designed it with a zoom factor of 0.5, retina vs HD size in mind. I thought adding a x2 zoom on the panel would be very 4k and HiDPI compliant.

So I struggled with that since I cannot get the VST frame fit the actual panel size. 
I stumbled uppon the following thread on Ctrlr.org forum and thanks to Mr.ToR I could get it right.
[export restricted instance + zoom panel problem](https://ctrlr.org/forums/topic/export-restricted-instance-hidden-menu-bar/)

@dobo365 Goodweather suggested a neat panel size setup in the following post : [create a uiSlider with for example 50 to 150 as value range...](https://ctrlr.org/forums/topic/export-restricted-instance-hidden-menu-bar/#post-98976)

Is Goodweather's zoom trick working only for standalone restricted apps or is it working for VST restricted instances as well?
I really need to have this right since I worked on 5 not-yet-completed panels and they are all designed with x1 or x2 size in mind for 4k compatibily.

Thanks!

**Comment by damiensellier** [comment]
I tried Goodweather's trick for the zoom. It's working only for standalone instances not as a VST.

Is it possible to resize the VST frame while loaded in the host ?

Which following property is the right one ?

```
panel:getCanvas():setSize(W,H)
panel:getPanelEditor():setProperty("uiPanelZoom",zoomFactor,true)
panel:getPanelEditor():setProperty("uiPanelCanvasRectangle",PosX,PosY,W,H,true)
```


Thanks ;)

**Comment by Godlike-Productions** [comment]
I doubt it, though there are plugins around that allow the frame to be resized, and older versions of ctrlr allowed the frame to be resized.

I think our best bet would be to change the size and reload the VST with the saved settings.  I'm not sure if the frame size is fixed when the restricted instance is created though.

**Reply by damiensellier** [reply]
Well I tried to assign a zoom button to the different properties.
The VST window size is "uiPanelCanvasRectangle"
It's setup when the panel is loading and you cannot modify it afterward.

I tried something like that:

```
zoomUI = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)

	local ZoomValue = panel:getModulatorByName("zoomTest"):getModulatorValue()

		if ZoomValue == 1 then

		panel:getCanvas():setSize(960,68)
		panel:getPanelEditor():setProperty("uiPanelZoom",0.5,true)
		panel:getPanelEditor():setProperty("uiPanelCanvasRectangle","0 0 960 68",true)		

		elseif ZoomValue == 0 then
		panel:getCanvas():setSize(1980,136)
		panel:getPanelEditor():setProperty("uiPanelZoom",1,true)
		panel:getPanelEditor():setProperty("uiPanelCanvasRectangle","0 0 1980 136", true)

		end -- endif 

end
```

Everything resized within the VST window but the frame doesn't change at all. So we are here for VST unless there's a size property I'm missing. I am thinking about making two different plugins, 1 exported for HD screens and a bigger one for Retina HiDPI 4K.. screens.

**Comment by damiensellier** [comment, marked as answer]
The only way to export VST with a zoom ratio other than 1.0 is what Mr.ToR suggested. I post it here as the solution : 

Let’s consider you have a panel with a size of **1910 x 1330** and you want it to be at **0.7** zoom, which makes 1337 x 931, so set your “uiPanelCanvasRectangle” to “**0 0 1337 931**” by entering the values manually and insert the following line to your ‘**onPanelLoad**’ method.

```
panel:getCanvas():setSize(1910,1330)
panel:getPanelEditor():setProperty("uiPanelZoom",0.7,true)
```

This will make your panel programmatically cropped but have it viewed at the correct size. The issue here is that host window size is set by the panel:getCanvas():getWidth() and getHeight() parameters, which does not change with zoom. That’s why we’re using a cropped panel. However, panel:getCanvas():getParentWidth() and getParentHeight() gives a size dependent on the zoom setting, and host window size should use those, even when if you still don’t have the ability to change the host window size on runtime.

Also, Ctrlr v5.3.198 on Mac, crops the window with 24 pixels from the bottom when you export a restricted instance, so to compensate that, set your “uiPanelCanvasRectangle” to “0 0 1337 955” adding 24 pixels to height.
